p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/tools/P2RepositoryDataManipulator.java

```diff
@@ -0,0 +1,193 @@
+			List<String> propertiesToRemove, Map<String, String> propertiesToAdd)
+			throws ProvisionException, URISyntaxException, IOException {
+
+		Consumer<IRepository<?>> modification = repo -> {
+			if (repositoryName != null) {
+				repo.setProperty(IRepository.PROP_NAME, repositoryName);
+			}
+			repo.setProperty(IRepository.PROP_COMPRESSED, Boolean.toString(repository.compress()));
+			if (propertiesToRemove != null) {
+				for (String key : propertiesToRemove) {
+					repo.setProperty(key, null);
+				}
+			}
+			if (propertiesToAdd != null) {
+				propertiesToAdd.forEach((key, value) -> {
+					repo.setProperty(key, value);
+				});
+			}
+		};
+
+		Repository location = repository.repositoryLocation();
+		P2RepositoryKind kind = repository.repositoryKind();
+		Path outputLocation = repository.destinationLocation();
+
+		if (kind.isArtifact()) {
+			modifyOutputRepository(manager.getArtifactRepository(location), outputLocation,
+					this::mirrorArtifactRepositoryData, r -> r.executeBatch(m -> modification.accept(r), null));
+		}
+		if (kind.isMetadata()) {
+			modifyOutputRepository(manager.getMetadataRepository(location), outputLocation,
+					this::mirrorMetadataRepository, r -> r.executeBatch(m -> modification.accept(r), null));
+		}
+		xzCompress(repository, outputLocation);
+	}
+
+	private <T, R extends IRepository<T>> void modifyOutputRepository(R repository, Path outputLocation,
+			BiFunction<R, Path, R> copyAction, Consumer<R> modification) throws ProvisionException {
+		if (outputLocation != null) {
+			R localRepository = copyAction.apply(repository, outputLocation);
+			modification.accept(localRepository);
+		} else if (repository.isModifiable()) {
+			modification.accept(repository);
+		} else {
+			throw new IllegalStateException(
+					"Repository is not modifable and no output location is defined: " + repository.getLocation());
+		}
+	}
+
+	/**
+	 * Creates a local mirror only of the <b>data</b> of the given
+	 * {@link IMetadataRepository}, i.e. its {@code artifacts.xml/jar/xml.xz} files.
+	 */
+	private IArtifactRepository mirrorArtifactRepositoryData(IArtifactRepository repository, Path targetLocation) {
+		try {
+			IArtifactRepository mirror = manager.createLocalArtifactRepository(targetLocation, repository.getName(),
+					repository.getProperties());
+			IStatus result = mirror.executeBatch(m -> {
+				// Copy artifacts to modifiable output repository
+				repository.query(ArtifactKeyQuery.ALL_KEYS, null).stream()
+						.map(a -> repository.getArtifactDescriptors(a)).forEach(a -> mirror.addDescriptors(a, null));
+				// Copy (potentially customized) mapping rules
+				if (repository instanceof SimpleArtifactRepository simpleSource
+						&& mirror instanceof SimpleArtifactRepository simpleTarget) {
+					simpleTarget.setRules(simpleSource.getRules()); // Copy artifact mapping rules
+				}
+			}, null);
+			assertNoError(result, targetLocation);
+			// TODO: remove repository from manager again?
```

Comment:
Should we make sure the source repository is again removed from the manager? or at least a potential (local) output/target repository?
I saw this in multiple occasions and think that it could make sense to keep the manager clean of such modified/worked on repos?

Reply:
See my comment above, when we can avoid adding it to the set of managed repos at the first place it would be better than removing them later on.

Reply:
Yes. Initially I also thought (and implemented it locally) that even the source repository is removed again, if it's a local repo, a remote repo is IMO still reasonable to cache.
But for the sake of simplicity I have now dropped that part as at least the use case in the platform-releng only involves a remote repo to modify.
If a cached local source repo makes trouble, it can still be enhanced later.